clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
```

Comment:
I think the API for poll() should **ideally** say the following:

if (fetch offset for any partition is out-of-range)
  throw OffsetOutOfRange exception
else {
  return data for all partitions who have data after the current fetch offset, before timeout
}

The user can easily know whether the offset is valid and whether there is data by looking at vale returned from poll().

I think the API would be more difficult to explain with your proposal, because I thought it would become the following:

if (there is OffsetOutOfRange exception for other partition in previous patch && 
    there is data for this partition from previous patch)
  return data
else if (fetch offset for any partition is out-of-range)
  throw OffsetOutOfRange exception
else {
  return data for all partitions who have data after the current fetch offset, before timeout
}

In this case, suppose user receives data for partition x from poll(), he can no longer decide if it is because the current offset is valid.

However, my concern is invalid, since value returned from poll() doesn't necessary come from the most recent fetchRequest.

Neither of us are strong on either solutions. I will update the patch as you suggested and move forward:)